src/number_theory/divisors.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+import data.nat.totient
+import number_theory.quadratic_reciprocity
+
+/-!
+# Divisor finsets
+
+This file defines sets of divisors of a natural number. This is particularly useful as background
+for defining Dirichlet convolution.
+
+## Main Definitions
+Let `n : ℕ`. All of the following definitions are in the `nat` namespace:
+ * `divisors n` is the `finset` of natural numbers that divide `n`.
+ * `proper_divisors n` is the `finset` of natural numbers that divide `n`, other than `n`.
+ * `divisors_antidiagonal n` is the `finset` of pairs `(x,y)` such that `x * y = n`.
+ * `perfect n` is true when the sum of `proper_divisors n` is `n`.
+
+## Implementation details
+ * All of `divisors 0`, `proper_divisors 0`, and `divisors_antidiagonal 0` are defined to be `∅`.
+
+## Tags
+divisors, perfect numbers
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+namespace nat
+variable (n : ℕ)
+
+/-- `divisors n` is the `finset` of divisors of `n`. As a special case, `divisors 0 = ∅`. -/
+def divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.Ico 1 (n + 1))
+
+/-- `proper_divisors n` is the `finset` of divisors of `n`, other than `n`.
+  As a special case, `divisors 0 = ∅`. -/
+def proper_divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.Ico 1 n)
+
+/-- `divisors_antidiagonal n` is the `finset` of pairs `(x,y)` such that `x * y = n`.
+  As a special case, `divisors_antidiagonal 0 = ∅`. -/
+def divisors_antidiagonal : finset (ℕ × ℕ) :=
+((finset.Ico 1 (n + 1)).product (finset.Ico 1 (n + 1))).filter (λ x, x.fst * x.snd = n)
+
+variable {n}
+
+lemma not_proper_self : ¬ n ∈ proper_divisors n :=
+begin
+  rw proper_divisors,
+  simp,
+end
+
+@[simp]
+lemma mem_proper_divisors {m : ℕ} : n ∈ proper_divisors m ↔ n ∣ m ∧ n < m :=
+begin
+  rw proper_divisors,
+  simp only [finset.Ico.mem, finset.mem_filter],
+  refine ⟨λ h, ⟨h.2, h.1.2⟩, λ h, ⟨⟨_, h.2⟩, h.1⟩⟩,
+  contrapose! h,
+  rintro ⟨c, rfl⟩,
+  simp [nat.le_zero_iff.1 (nat.lt_add_one_iff.1 h)]
+end
+
+lemma divisors_eq_proper_divisors_insert_self (posn : 0 < n) :
+  divisors n = has_insert.insert n (proper_divisors n) :=
+by rw [divisors, proper_divisors, finset.Ico.succ_top posn,
+       finset.filter_insert, if_pos (dvd_refl n)]
+
+@[simp]
+lemma mem_divisors {m : ℕ} :
+  n ∈ divisors m ↔ n ∣ m ∧ m ≠ 0:=
+begin
+  cases m, {simp [divisors]},
```

Comment:
Could you make sure this file matches the [style guidelines](https://leanprover-community.github.io/contribute/style.html)? I see quite a few cases of inconsistent bracket placement, spacing around symbols and indentation, which makes it more annoying to read.